### PR TITLE
Animate hitscan shot and add various effects

### LINKED ISSUE #188: Animate the hitscan shot

### PR DESCRIPTION
**Pull request by kb173**
In addition to the animation, the color was changed to the main player color, and some code was added to PostProcess (e.g. global access to camera shake) and used in the hitscan shot.

https://user-images.githubusercontent.com/18697097/147004457-e9122f80-37b6-4ccb-a05e-44a28acb4869.mp4

closes #188

### CODE REVIEW

File: recursio-client/Global/PostProcess.gd [hunk 1]
```diff
@@ -3,53 +3,53 @@ extends Node
...
 
 
-func remove_animation(property: NodePath):
-	_tween.remove(self, property)
+func animate_property(property: String, initial_val, final_val, duration: float, trans_type = Tween.TRANS_LINEAR, ease_type = Tween.EASE_OUT):
+	var tween = Tween.new()
+	add_child(tween)
+	tween.interpolate_property(self, property, initial_val, final_val, duration, trans_type, ease_type)
+	tween.start()
+	tween.connect("tween_all_completed", tween, "queue_free")
```
Comment (AmonShokhinAhmed) [inline recursio-client/Global/PostProcess.gd:31]: I'm not sure it matters, but isn't very inefficient to always instance a new tween and then free it on completion?

File: recursio-client/Global/PostProcess.gd [hunk 2]
```diff
@@ -3,53 +3,53 @@ extends Node
...
-#glitch
-var glitch :bool = false setget set_glitch, get_glitch
-# vignette
-var vignette :bool = false setget set_vignette, get_vignette
-var vignette_softness: float = 1.0 setget set_vignette_softness, get_vignette_softness
+# Glitch
+var glitch :bool = false setget set_glitch
+
+# Vignette
+var vignette :bool = false setget set_vignette
```
Comment (AmonShokhinAhmed) [inline recursio-client/Global/PostProcess.gd:11]: I mean I know they aren't necessary yet, but why did you remove all getter?

File: Shared/Actions/Shots/HitscanShot.gd [hunk 3]
```diff
@@ -2,10 +2,18 @@ extends RayCast
 
 onready var _max_time = Constants.get_value("hitscan", "max_time")
 onready var _bullet_range = Constants.get_value("hitscan", "range")
+onready var _camera_shake_amount = Constants.get_value("vfx","shoot_camera_shake_amount")
+onready var _camera_shake_speed  = Constants.get_value("vfx","shoot_camera_shake_speed")
+onready var _camera_shake_duration  = 0.2
```
Comment (AmonShokhinAhmed) [inline Shared/Actions/Shots/HitscanShot.gd:7]: Should we not add this to constants as well?

File: Shared/Actions/Shots/HitscanShot.gd [hunk 4]
```diff
@@ -18,19 +26,23 @@ func initialize(owning_player) -> void:
...
 		color_manager.color_object_by_property(color_name, $Visualisation.material_override, "albedo_color")
+		color_manager.color_object_by_property(color_name, $Visualisation.material_override, "emission")
 		color_manager.color_object_by_property(color_name, $HitPoint/FrontParticles.material_override, "emission")
 		color_manager.color_object_by_property(color_name, $HitPoint/BackParticles.material_override, "emission")
 
 	cast_to = Vector3(0,0,-_bullet_range)
 	_owning_player = owning_player
 	add_exception(owning_player.get_body())
+	
+	PostProcess.animate_property("chromatic_ab_strength", 0.5, 0, _max_time)
```
Comment (AmonShokhinAhmed) [inline Shared/Actions/Shots/HitscanShot.gd:44]: This now also triggers those effects if any other character than the player spawns a hitscan shot. I don't think we want this.

File: Shared/Actions/Shots/HitscanShot.gd [hunk 5]
```diff
@@ -43,15 +55,17 @@ func _physics_process(delta):
...
 func handle_hit(collider):
 	var character = collider.get_parent()
 	Logger.debug("hit collider: %s" %[collider.get_class()] , "HitscanShot")
 	var collision_point = get_collision_point()
 	var distance = (collision_point- global_transform.origin).length()
-	$Visualisation.scale.y = distance*0.5
-	$Visualisation.transform.origin.z = -distance*0.5
+	_current_range = distance
 
 	if _first_frame:
```
Comment (AmonShokhinAhmed) [inline Shared/Actions/Shots/HitscanShot.gd:70]: We actually only update collisions during the first frame for some reason. ( I think this is also a relict from the prototype). I think we should change that so we don't have visual hits that don't trigger anything.

**Review by AmonShokhinAhmed — CHANGES_REQUESTED** [review]
The visuals work very well IMO. Just a few clarifications are needed.

File: recursio-client/Util/LerpedFollow.gd [hunk 6]
```diff
@@ -27,6 +31,12 @@ func start_following():
 	set_physics_process(true)
 
 
+func shake(amount: float, speed: float, duration: float):
+	start_shake(amount, speed)
+	yield(get_tree().create_timer(duration), "timeout")
```
[the remark this answers is not in the document]
Comment (kb173) [reply recursio-client/Util/LerpedFollow.gd:36]: Good point, it might not. I'll try to add a counter to see if anyone else is shaking the camera at the moment, and prevent stop_shake if so.

File: recursio-client/Global/PostProcess.gd (hunk 2 above)
Comment (kb173) [reply recursio-client/Global/PostProcess.gd:11]: I didn't see any possible reason why we'd want to do something specific in a getter other than returning the variable, or is there something you had in mind?

File: Shared/Actions/Shots/HitscanShot.gd (hunk 4 above)
Comment (kb173) [reply Shared/Actions/Shots/HitscanShot.gd:44]: Don't we? I found the effect to be pretty cool, it makes combat very lively, and it thematically kind of makes sense that the other character shooting causes this "glitch" too. But if the consensus is that we don't want it, I can change it. (@incredibleLeitman @DavidPgl)

File: Shared/Actions/Shots/HitscanShot.gd (hunk 5 above)
Comment (kb173) [reply Shared/Actions/Shots/HitscanShot.gd:70]: Oh really? damn, yeah I'll change that

File: recursio-client/Global/PostProcess.gd (hunk 2 above)
Comment (AmonShokhinAhmed) [reply recursio-client/Global/PostProcess.gd:11]: Not sure, just thought we sometime might want to animate from the current value to another value or something similar. And as of now we have no way to do that. But yea might as well only add it when it's needed.

File: Shared/Actions/Shots/HitscanShot.gd (hunk 4 above)
Comment (AmonShokhinAhmed) [reply Shared/Actions/Shots/HitscanShot.gd:44]: I mean, if we compare it to other games, your really only get camera shake etc. when something directly concerns you. A random ghost shooting in the corner opposite from the player triggering the shake might really confuse the player.

Comment (AmonShokhinAhmed) [reply Shared/Actions/Shots/HitscanShot.gd:44]: But yeah its hard to say, without testing it during a real game.

Comment (kb173) [reply Shared/Actions/Shots/HitscanShot.gd:44]: True, we might want to deactivate it for ghosts. But I think it'd be cool to have at least for the other active player. Should I leave it as is for now and open a 'Needs Testing' issue for it after it's merged?

File: recursio-client/Global/PostProcess.gd (hunk 2 above)
Comment (kb173) [reply recursio-client/Global/PostProcess.gd:11]: Maybe there's a misunderstanding: omitting the getter doesn't delete it, that has to be done explicitly. It just uses the default getter (the one that was previously manually implemented).

File: Shared/Actions/Shots/HitscanShot.gd (hunk 4 above)
Comment (DavidPgl) [reply Shared/Actions/Shots/HitscanShot.gd:44]: Maybe only do it when the player shoots or gets hit (enemy or ghost)?

Comment (kb173) [reply Shared/Actions/Shots/HitscanShot.gd:44]: When the player gets hit, we have the more exaggerated death effect anyways

Comment (AmonShokhinAhmed) [reply Shared/Actions/Shots/HitscanShot.gd:44]: > True, we might want to deactivate it for ghosts. But I think it'd be cool to have at least for the other active player. Should I leave it as is for now and open a 'Needs Testing' issue for it after it's merged?

Maybe extract the triggering to the character base class so that if we decide to disable it for other characters, we can just flip a boolean or something. But yea should be fine with a needs testing issue.

File: recursio-client/Global/PostProcess.gd (hunk 2 above)
Comment (AmonShokhinAhmed) [reply recursio-client/Global/PostProcess.gd:11]: ohhhh yeah there was a misunderstanding then, my bad.

File: Shared/Actions/Shots/HitscanShot.gd (hunk 4 above)
Comment (kb173) [reply Shared/Actions/Shots/HitscanShot.gd:44]: Sounds good :+1:

**Comment by kb173** [comment]
Should be fixed now

**Comment by AmonShokhinAhmed** [comment]
Quickly fixed a small isssue but seems great otherwise.
